### ISSUE TITLE
AspNetCore binaries introduced into the VMR

### ISSUE THREAD

**Issue by oleksandr-didyk** [dotnet/source-build#4161]
The following binaries from the AspNetCore repo were introduced into the VMR by [sync](https://dev.azure.com/dnceng/internal/_git/dotnet-dotnet/commit/2109df3caf1bae2d005485f7dc4e3f551844ca9f?path=/src/templating/docs/Using-Group-Identity.md)

```
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha1/00/62/9c35a688e05a88b1cda684fb9d5e73f000a1
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha1/00/73/97d44ed1872fdc6ed31360190f81814e2c64
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha1/00/7f/15240b5751c537c40e77abb4e89eeaaa8847
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha1/01/24/65bf940d12375dc47387b754ccf9d6b92180
...
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha512/be/a2/7fa40521f7914b73f1bfdfec4faa548182294f9dab957f6117390099ce16490970d7b7fe40f261ff3b353a89685bdc2364e613c9ae1f29a8950f95d379cf
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha512/c7/50/966eeae56cde498c6fa3b9e9ee35ae0a75e7417572e364ca79ea71c54874ceb1bfcc2360d8886ccbb5118734ccb274f19f7c621df9582338cadc96d3a4b0
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha512/c9/9f/8c51b9f07f748735640a2725a1f3c8a2d876ee1c21509c061379e3b535bf12c84896012122e5d414ca0b770a51d6b40f4ccbd450018923bebb19d162bc5e
src/aspnetcore/src/submodules/Node-Externals/cache/_cacache/content-v2/sha512/ca/6e/71d5c8f899400c06e9a6c71464238a46e5586dd5032c25ec7018ff39e3b4d9d2ca31f5f08b0c7c3d2821fd73514ee5df46a29be4689b0120e9132b192c77
```

The first build where these pop-up: https://dev.azure.com/dnceng/internal/_build/results?buildId=2378459&view=results

**Comment by oleksandr-didyk** [comment]
Seems to be caused by https://github.com/dotnet/aspnetcore/pull/53975

**Comment by mthalman** [comment]
Well this isn't good. @wtgodbe - Binary files aren't allowed for source build in the VMR.

**Comment by NikolaMilosavljevic** [comment]
Is there an alternative solution to checking in the cache? If these are binaries and prebuilt minified json files, can they be built from sources?

There is the orthogonal issue (https://github.com/dotnet/source-build/issues/4149) that was caused by introduction of these cache files. Scanning the whole cache is timing out - some quick testing show that with increased timeout it would take more than 24 hours to do license scan of the whole cache. Is there a way to figure out which cache files contain licenses, so that we can exclude the others from scanning?

**Comment by wtgodbe** [comment]
> Is there an alternative solution to checking in the cache? If these are binaries and prebuilt minified json files, can they be built from sources?

The alternative solution was to have our source build partners build our node components online as a pre-build step. The checked-in cache was a solution we came with explicitly for our source-build partners, to avoid having to restore NPM packages over the network during offline sourcebuild.

**Comment by javiercn** [comment]
> Is there an alternative solution to checking in the cache? If these are binaries and prebuilt minified json files, can they be built from sources?

The vast majority of our packages are for the build tools that our JS packages use to build our JS, it's not realistic that we source build 1.4K+ node dependencies.

> There is the orthogonal issue (#4149) that was caused by introduction of these cache files. Scanning the whole cache is timing out - some quick testing show that with increased timeout it would take more than 24 hours to do license scan of the whole cache. Is there a way to figure out which cache files contain licenses, so that we can exclude the others from scanning?

The cache should be treated as an opaque blob and skipped. Its contents are described by the `package-lock.json` file in the repo and can't change without the build failing (the cache is integrity validated and can't go out of sync).

License scanning should happen against the packages described in package-lock.json since the contents of the cache are an exact copy of the packages described in there.

As @wtgodbe this was an agreement that we reached with our source build partners.

**Comment by mthalman** [comment]
Pinging @dotnet/distro-maintainers for awareness.

**Comment by NikolaMilosavljevic** [comment]
I see. So, these are various npm packages that are needed for building aspnetcore. I presume there is no way to build those packages from source?

**Comment by mthalman** [comment]
@omajid - Can you provide your input of what Red Hat would allow for binaries? What kind of binaries would these need to be in order to be acceptable for Red Hat?

@javiercn - At a minimum, I think we need to have an understanding of the content of these files and whether they're actually required or not.

**Comment by dviererbe** [comment]
Note: Not a complete review.

I looked into a bunch of these files and (as @omajid already pointed out during the last security partners meeting) they appear to be gzip archives containing the source of the npm packages or metadata files (mostly in json format). Therefore in general they should be fine to include in Ubuntu, because technically we are still building from source, but we should still screen for minifyed js/ts without a non-minifyed counterpart & executable native binaries contained within them.

One aspect that is a bit annoying, but I am fine with is, that it would be a nightmare to write a patch for any of the gzip archives, because of the compression and strict hash matching.

**Comment by omajid** [comment]
With my Fedora/RHEL distro-maintainer hat on, I want to avoid including binaries in our source trees as much as possible. There's a couple of reasons for that, including licensing, being able to modify/patch things as needed, the chance we are using something we can't build, and the risk of accidentally shipping that prebuilt binary to users.

Licensing is the hard blocker: I am not allowed to include anything with a non-open license in Fedora/RHEL. If there is a chance the binary is not under an open source license, I have to pre-delete it from the .NET source archive before I can add it to the distro.

If the binary is known to be under an open source license, I can use the source archive as-is and just have to delete the specific binary before the actual .NET build phase. (There's some limited exception to this rule, which comes in handy when I have to bootstrap .NET).

When we had our discussion with the ASP.NET Core team, I was fully on board with using the cache. However, I had een looking at [how Fedora processed npm dependnecies](https://src.fedoraproject.org/rpms/nodejs-packaging/blob/rawhide/f/nodejs-packaging-bundler) and I completely misunderstood what the npm cache looks like. I had expected it to match the contents of `node_modules/` directory, and did not expect this content-addressable cache (`cacache`) system, which heavily uses tarballs as the internal system for sources and binaries.

IMO, the primary problem with these tar files isn't that they are binaries. The VMR includes many non-text (aka binary) files. There are several hundred `.png` files in the VMR. I think the concerning thing is that these tar files contains other files, and we need to look into those for several reasons, such as license and binary scans.

As a concrete example, the binary scan doesn't look inside the tar files, and doesn't catch that one file (`_cacache/content-v2/sha512/be/a2/7fa40521f7914b73f1bfdfec4faa548182294f9dab957f6117390099ce16490970d7b7fe40f261ff3b353a89685bdc2364e613c9ae1f29a8950f95d379cf`) contains a binary (`package/rollup.darwin-arm64.node`) that I do need to flag and remove at distro build time. But if I do, that breaks npm's cache verification, right?

@dviererbe has pointed out that these compressed files also make it harder to quickly fix a piece of code if we, say, find a security issue that's actively being exploited.

To put it another way, I have two primary concerns:

- Is there some way to confirm that everything in this tar-based-cache system is fully open source?
- Can we (reliably) remove actual binaries that contain executable code (eg, exe files, shared libraries, dll) from within the cache?

And sorry about the delay in getting back. Node.js isn't my area of expertise, so I had to spend some time digging into things a bit.

**Comment by omajid** [comment]
https://github.com/dotnet/aspnetcore/issues/54393 (and the possible fix https://github.com/rollup/rollup/pull/5346) has confirmed that the Node-Externals cache contains prebuilt native binaries. That's a problem for us distro-maintainers.
